src/main/java/com/minecolonies/coremod/colony/buildings/inventory/ItemStackHandlerWithIndex.java

```diff
@@ -0,0 +1,333 @@
+package com.minecolonies.coremod.colony.buildings.inventory;
+
+import com.minecolonies.api.crafting.ItemStorage;
+import com.minecolonies.api.util.ItemStackUtils;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.NonNullList;
+import net.minecraftforge.items.ItemStackHandler;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * ItemStackHandler which also keeps an index of its contents.
+ * It also can push any changes to a meta(upper) index, to combine multiple handlers.
+ *
+ * @param <Value> The type of the meta index, not required.
+ */
+public class ItemStackHandlerWithIndex<Value> extends ItemStackHandler
+{
+    /**
+     * Index for stored Items and their slot.
+     */
+    @NotNull
+    private ItemStorageIndex<Integer> index;
+
+    /**
+     * A meta index which is to be updated. Does not have to be set.
+     */
+    @Nullable
+    private ItemStorageIndex<Value> meta;
```

Comment:
Not needed!
The implementation should fallback onto itemstacks with a custom hashing mechanism this is not fast enough and does not cover all items, since you seem to disregard NBT, which will break in 1.13 then anyway.